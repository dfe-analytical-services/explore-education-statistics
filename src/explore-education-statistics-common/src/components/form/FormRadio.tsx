import useMounted from '@common/hooks/useMounted';
import styles from '@common/components/form/FormRadio.module.scss';
import classNames from 'classnames';
import React, {
  ChangeEvent,
  FocusEventHandler,
  memo,
  ReactNode,
  Ref,
} from 'react';

export type OtherRadioChangeProps = Pick<FormRadioProps, 'label'>;

export type RadioChangeEventHandler = (
  event: ChangeEvent<HTMLInputElement>,
  radioProps: OtherRadioChangeProps,
) => void;

export interface FormRadioProps {
  checked?: boolean;
  className?: string;
  conditional?: ReactNode;
  defaultChecked?: boolean;
  disabled?: boolean;
  divider?: string;
  hiddenConditional?: boolean;
  hint?: string | ReactNode;
  hintSmall?: boolean;
  inlineHint?: boolean;
  id: string;
  inputRef?: Ref<HTMLInputElement>;
  label: string;
  labelClassName?: string;
  name: string;
  value: string;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: RadioChangeEventHandler;
}

const FormRadio = ({
  checked,
  className,
  conditional,
  defaultChecked,
  disabled = false,
  divider,
  hiddenConditional,
  hint,
  hintSmall = false,
  id,
  inlineHint,
  inputRef,
  label,
  labelClassName,
  name,
  value,
  onBlur,
  onChange,
}: FormRadioProps) => {
  const { onMounted } = useMounted(undefined, false);

  /* eslint-disable jsx-a11y/role-supports-aria-props */
  return (
    <>
      <div
        className={classNames('govuk-radios__item', className, {
          [styles.inlineHint]: inlineHint,
        })}
        data-testid={`Radio item for ${label}`}
      >
        <input
          aria-describedby={hint ? `${id}-item-hint` : undefined}
          aria-controls={onMounted(
            conditional ? `${id}-conditional` : undefined,
          )}
          aria-expanded={onMounted(conditional ? checked : undefined)}
          className="govuk-radios__input"
          checked={checked}
          defaultChecked={defaultChecked}
          disabled={disabled}
          id={id}
          name={name}
          onBlur={onBlur}
          onChange={event => {
            if (onChange) {
              onChange(event, { label });
            }
          }}
          ref={inputRef}
          type="radio"
          value={value}
        />
        <label
          className={classNames(
            'govuk-label govuk-radios__label',
            labelClassName,
          )}
          htmlFor={id}
        >
          {label}
        </label>
        {hint && (
          <span
            id={`${id}-item-hint`}
            className={classNames('govuk-hint govuk-radios__hint', {
              'govuk-!-font-size-14': hintSmall,
            })}
          >
            {hint}
          </span>
        )}
      </div>

      {conditional && (
        <div
          className={classNames('govuk-radios__conditional', {
            'govuk-radios__conditional--hidden': !checked || hiddenConditional,
          })}
          id={`${id}-conditional`}
        >
          {conditional}
        </div>
      )}

      {divider && <div className={styles.divider}>{divider}</div>}
    </>
  );
};

export default memo(FormRadio);
