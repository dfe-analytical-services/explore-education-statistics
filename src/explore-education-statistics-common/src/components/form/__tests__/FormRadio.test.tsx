import { render } from '@testing-library/react';
import React from 'react';
import FormRadio from '../FormRadio';

describe('FormRadio', () => {
  test('renders correctly with required props', () => {
    const { container } = render(
      <FormRadio name="test" id="test-radio" label="Test radio" value="true" />,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly when `hint` is provided', () => {
    const { container, getByLabelText, getByText } = render(
      <FormRadio
        name="test"
        hint="Click me to proceed"
        id="test-radio"
        label="Test radio"
        value="true"
      />,
    );

    expect(getByText('Click me to proceed')).toHaveAttribute(
      'id',
      'test-radio-item-hint',
    );
    expect(getByLabelText('Test radio')).toHaveAttribute(
      'aria-describedby',
      'test-radio-item-hint',
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders conditional content when checked', () => {
    const { container, getByText } = render(
      <FormRadio
        name="test"
        id="test-radio"
        label="Test radio"
        value="true"
        checked
        onChange={() => {}}
        conditional={<p>The conditional content</p>}
      />,
    );

    expect(getByText('The conditional content').parentElement).not.toHaveClass(
      'govuk-radios__conditional--hidden',
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('does not render conditional content when not checked', () => {
    const { container, getByText } = render(
      <FormRadio
        name="test"
        id="test-radio"
        label="Test radio"
        value="true"
        conditional={<p>The conditional content</p>}
      />,
    );

    expect(getByText('The conditional content').parentElement).toHaveClass(
      'govuk-radios__conditional--hidden',
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
});
