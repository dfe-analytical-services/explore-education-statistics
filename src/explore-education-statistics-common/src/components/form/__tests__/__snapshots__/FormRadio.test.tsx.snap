// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRadio does not render conditional content when not checked 1`] = `
<div class="govuk-radios__item"
     data-testid="Radio item for Test radio"
>
  <input class="govuk-radios__input"
         id="test-radio"
         name="test"
         type="radio"
         value="true"
         aria-controls="test-radio-conditional"
  >
  <label class="govuk-label govuk-radios__label"
         for="test-radio"
  >
    Test radio
  </label>
</div>
<div class="govuk-radios__conditional govuk-radios__conditional--hidden"
     id="test-radio-conditional"
>
  <p>
    The conditional content
  </p>
</div>
`;

exports[`FormRadio renders conditional content when checked 1`] = `
<div class="govuk-radios__item"
     data-testid="Radio item for Test radio"
>
  <input class="govuk-radios__input"
         id="test-radio"
         name="test"
         type="radio"
         value="true"
         checked
         aria-controls="test-radio-conditional"
         aria-expanded="true"
  >
  <label class="govuk-label govuk-radios__label"
         for="test-radio"
  >
    Test radio
  </label>
</div>
<div class="govuk-radios__conditional"
     id="test-radio-conditional"
>
  <p>
    The conditional content
  </p>
</div>
`;

exports[`FormRadio renders correctly when \`hint\` is provided 1`] = `
<div class="govuk-radios__item"
     data-testid="Radio item for Test radio"
>
  <input aria-describedby="test-radio-item-hint"
         class="govuk-radios__input"
         id="test-radio"
         name="test"
         type="radio"
         value="true"
  >
  <label class="govuk-label govuk-radios__label"
         for="test-radio"
  >
    Test radio
  </label>
  <div id="test-radio-item-hint"
       class="govuk-hint govuk-radios__hint"
  >
    Click me to proceed
  </div>
</div>
`;

exports[`FormRadio renders correctly with required props 1`] = `
<div class="govuk-radios__item"
     data-testid="Radio item for Test radio"
>
  <input class="govuk-radios__input"
         id="test-radio"
         name="test"
         type="radio"
         value="true"
  >
  <label class="govuk-label govuk-radios__label"
         for="test-radio"
  >
    Test radio
  </label>
</div>
`;
